src/operator/src/statement.rs

```diff
@@ -429,6 +432,115 @@ impl StatementExecutor {
                 table_name: table_ref.to_string(),
             })
     }
+
+    async fn plan_exec_inner(&self, stmt: QueryStatement, query_ctx: QueryContextRef, timeout: Option<Duration>) -> Result<Output> {
+        let exec_query = |stmt: QueryStatement, ctx: QueryContextRef| {
+            async move {
+                let plan = self.plan(stmt, ctx.clone()).await?;
+                self.query_engine.execute(plan, ctx).await.context(ExecLogicalPlanSnafu)
+            }
+        };
+        if let Some(t) = timeout {
+            let start = tokio::time::Instant::now();
+            let timeout_future = tokio::time::sleep(t);
+            tokio::select! {
+                result = exec_query(stmt, query_ctx.clone()) => {
+                    let new_timeout = t - start.elapsed();
+                    return Ok(attach_timeout_stream(result?, new_timeout));
+                }
+                _ = timeout_future => {
+                    return StatementTimeoutSnafu.fail()
+                }
+            }
+        }
+        exec_query(stmt, query_ctx.clone()).await
+    }
+}
+
+fn attach_timeout_stream(output: Output, timeout: Duration) -> Output {
+    match output.data {
+        OutputData::AffectedRows(_) | OutputData::RecordBatches(_) => output,
+        OutputData::Stream(stream) => {
+            let stream = TimeoutStream::new(stream, timeout);
+            Output::new(OutputData::Stream(Box::pin(stream)), output.meta)
+        }
+    }
+}
+
+struct TimeoutStream {
+    stream: SendableRecordBatchStream,
+    timeout_duration: Duration,
+    last_poll: Option<tokio::time::Instant>,
+}
+
+impl TimeoutStream
+{
+    fn new(stream: SendableRecordBatchStream, timeout_duration: Duration) -> Self {
+        TimeoutStream {
+            stream,
+            timeout_duration,
+            last_poll: Some(tokio::time::Instant::now()),
+        }
+    }
+}
+
+impl RecordBatchStream for TimeoutStream {
+    fn name(&self) -> &str {
+        self.stream.name()
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.stream.schema()
+    }
+
+    fn output_ordering(&self) -> Option<&[OrderOption]> {
+        self.stream.output_ordering()
+    }
+
+    fn metrics(&self) -> Option<RecordBatchMetrics> {
+        self.stream.metrics()
+    }
+}
+
+impl Stream for TimeoutStream
+{
+    type Item = common_recordbatch::error::Result<RecordBatch>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        if let Some(last_poll) = this.last_poll {
+            if last_poll.elapsed() >= this.timeout_duration {
```

Comment:
a small issue: if a waker is send back after first pending poll and result is not polled again until waker signals ready, but its probably ok because client wont see this.